linker/Linker.Steps/MarkStep.cs

```diff
@@ -541,13 +572,23 @@ void ProcessLazyAttributes ()
 					continue;
 				}
 
-				// If an attribute's module has not been marked after processing all types in all assemblies and the attribute itself has not been marked,
-				// then surely nothing is using this attribute and there is no need to mark it
-				if (!Annotations.IsMarked (resolved.Module) && !Annotations.IsMarked (customAttribute.AttributeType))
+				if (!ShouldMarkTopLevelCustomAttribute (customAttribute, resolved)) {
+					skippedItems.Add (customAttribute);
 					continue;
+				}
 
 				MarkCustomAttribute (customAttribute);
 			}
+
+			// requeue the items we skipped in case we need to make another pass
+			foreach (var item in skippedItems)
+				_topLevelAttributes.Enqueue (item);
+
+			// If we have the same number of items afterwards, then we didn't do any additional marking
+			if (startingQueueCount == _topLevelAttributes.Count)
```

Comment:
I am not certain this will always work. Could the queue have an item removed and added at the same time?

Reply:
You're right.  `MarkCustomAttribute` could call  `ShouldParseMethodBody`, which could call `ResolveAssembly`.  And if that happened `MarkAssembly` would be called and new assembly or module attributes could be enqueued.

I will come up with at test case and switch to a more robust way of tracking if anything new was marked.

Reply:
I tried to write a test case and I wasn't able to figure out a way to cause a problem.  We call `LazyMarkCustomAttributes` on all assemblies returned by `_context.GetAssemblies()` before the first call to `ProcessAllQueues` even takes place. 

To cause an issue, you'd have to get the `ResolveAssembly` call in `ShouldParseMethodBody` to be called on an assembly that was not returned by `_context.GetAssemblies()`

I'm not sure how you'd do that and I don't think it's worth the effort to figure out.    So I will not be including an additional test case.

There was a straightward way to make the return value more robust, so I will still update the PR with that change.